### PR TITLE
Implement custom placement of modules in Mesos cluster

### PR DESCRIPTION
**Pull request by ericbottard**
Fixes https://github.com/spring-cloud/spring-cloud-deployer-mesos/issues/38

### CODE REVIEW

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncher.java [hunk 1]
```diff
@@ -244,5 +249,13 @@ private Double deduceCpus(AppDeploymentRequest request) {
 		return override != null ? Double.valueOf(override) : properties.getCpu();
 	}
 
+	private Collection<Constraint> deduceConstraints(AppDeploymentRequest request) {
+		Set<Constraint> requestSpecific = StringUtils.commaDelimitedListToSet(request.getDeploymentProperties().get("spring.cloud.deployer.chronos.constraints"))
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncher.java]: Shouldn't this property name be `spring.cloud.deployer.mesos.chronos.constraints`? 
Would be better to create a constant on the `prefix`.

Comment (ericbottard) [reply src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncher.java]: This is a known inconsistency (see https://github.com/spring-cloud/spring-cloud-deployer-mesos/issues/41) both for Chronos and Marathon. I chose to be consistent with how it's currently done, but definitely weird. cc @trisberg

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java [hunk 2]
```diff
@@ -292,6 +297,14 @@ private String extractGroupId(String appId) {
 		return groupId;
 	}
 
+	private Collection<Constraint> deduceConstraints(AppDeploymentRequest request) {
+		Set<Constraint> requestSpecific = StringUtils.commaDelimitedListToSet(request.getDeploymentProperties().get("spring.cloud.deployer.marathon.constraints"))
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java]: Same here as well. The property name prefix is missing `mesos`?

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java [hunk 3]
```diff
@@ -292,6 +297,14 @@ private String extractGroupId(String appId) {
...
+			.stream().map(Constraint::new).collect(Collectors.toSet());
+		Set<Constraint> result = new HashSet<>(properties.getConstraints());
+		result.addAll(requestSpecific);
+		return result;
+	}
+
 	private Collection<String> deduceUris(AppDeploymentRequest request) {
 		Set<String> additional = StringUtils.commaDelimitedListToSet(request.getDeploymentProperties().get("spring.cloud.deployer.marathon.uris"));
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java]: I see this issue in some other places as well. I think we definitely need a constant for the `prefix` and reuse everywhere the properties are used.

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncherProperties.java [hunk 4]
```diff
@@ -61,6 +65,11 @@
 	 */
 	private String ownerName;
 
+	/**
+	 * A set of constraints to apply to any deployed app, as a comma separated set of (field operator param?) triplets.
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncherProperties.java]: `deployed app` can be changed to `launched task`?

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java [hunk 5]
```diff
@@ -292,6 +297,14 @@ private String extractGroupId(String appId) {
 		return groupId;
 	}
 
+	private Collection<Constraint> deduceConstraints(AppDeploymentRequest request) {
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java:300]: The method `deduceConstraints` can possibly refactored to use `AppDeploymentRequest` and `DeploymentProperty key` and thereby can be used by both the app deployer and chronos task launcher.

Comment (ericbottard) [reply src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java:300]: Not the same `properties` class :(

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncherProperties.java (hunk 4 above)
Comment (ericbottard) [reply src/main/java/org/springframework/cloud/deployer/spi/mesos/chronos/ChronosTaskLauncherProperties.java]: Could be, but I chose not to, as this is really about placing an app/container (that is a "task" in our language, but...)

I'm fine with either way

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java (hunk 5 above)
Comment (ilayaperumalg) [reply src/main/java/org/springframework/cloud/deployer/spi/mesos/marathon/MarathonAppDeployer.java:300]: oh yeah, I missed that ;)

File: src/main/java/org/springframework/cloud/deployer/spi/mesos/constraints/Constraint.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+	private static final Pattern PARSE_REGEX = Pattern.compile("(?<field>[^ ]+) (?<op>[^ ]+)( (?<param>.+))?");
+
+	private final String field;
+
+	private final String operator;
+
+	private final String parameter;
+
+	public Constraint(String raw) {
```
Comment (ilayaperumalg) [inline src/main/java/org/springframework/cloud/deployer/spi/mesos/constraints/Constraint.java:43]: would be good to have some unit tests on the Constraint construction etc.,

Comment (ericbottard) [reply src/main/java/org/springframework/cloud/deployer/spi/mesos/constraints/Constraint.java:43]: will add

**Comment by ilayaperumalg** [comment]
This PR needs to be updated based on the `prefix` changes https://github.com/spring-cloud/spring-cloud-deployer-mesos/pull/43

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Squashed, rebased and merged as 07fc90f

**Pull request closed without merging**